src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp

```diff
@@ -467,6 +1111,117 @@ void four_index_constraint(
     }
   }
 }
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+tnsr::a<DataType, SpatialDim, Frame> f_constraint(
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_function,
+    const tnsr::ia<DataType, SpatialDim, Frame>& d_gauge_function,
+    const tnsr::a<DataType, SpatialDim, Frame>& spacetime_normal_one_form,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& d_pi,
+    const tnsr::ijaa<DataType, SpatialDim, Frame>& d_phi,
+    const Scalar<DataType>& gamma2,
+    const tnsr::iaa<DataType, SpatialDim, Frame>&
+        three_index_constraint) noexcept {
+  auto constraint =
+      make_with_value<tnsr::a<DataType, SpatialDim, Frame>>(pi, 0.0);
+  f_constraint<SpatialDim, Frame, DataType>(
+      &constraint, gauge_function, d_gauge_function, spacetime_normal_one_form,
+      spacetime_normal_vector, inverse_spatial_metric, inverse_spacetime_metric,
+      pi, phi, d_pi, d_phi, gamma2, three_index_constraint);
+  return constraint;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void f_constraint(
+    gsl::not_null<tnsr::a<DataType, SpatialDim, Frame>*> constraint,
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_function,
+    const tnsr::ia<DataType, SpatialDim, Frame>& d_gauge_function,
+    const tnsr::a<DataType, SpatialDim, Frame>& spacetime_normal_one_form,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& d_pi,
+    const tnsr::ijaa<DataType, SpatialDim, Frame>& d_phi,
+    const Scalar<DataType>& gamma2,
+    const tnsr::iaa<DataType, SpatialDim, Frame>&
+        three_index_constraint) noexcept {
+  if (get_size(get<0>(*constraint)) != get_size(get<0, 0>(pi))) {
+    *constraint = tnsr::a<DataType, SpatialDim, Frame>{get<0, 0>(pi)};
```

Comment:
Could you please construct the tensor with `..., Frame>(get_size(get<0, 0>(pi)), 0.0)` and then remove the `std::fill` below?

Reply:
`*constraint = tnsr::a<DataType, SpatialDim, Frame>(get_size(get<0, 0>(pi)), 0.0);` does not compile. 

The compiler error is as follows:
~~~~
error: no matching constructor for initialization of 'tnsr::a<double, 1UL, Grid>' (aka 'Tensor<double, brigand::list<brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<1, UpLo::Lo, Frame::Grid, IndexType::Spacetime> > >')
        tnsr::a<DataType, SpatialDim, Frame>(get_size(get<0, 0>(pi)), 0.0);
~~~~

Don't I need to loop over tensor components at set those individually to zero, which is basically what the `std::fill` was doing anyway? If not, can you see what's wrong with the way I implemented your suggestion?

Reply:
Yea, that's because it's a double. Never mind this will get taken care of during optimization.